### PR TITLE
Add info on the user responsible for adding a variant to a ClinVar submission + add/remove events

### PR DESCRIPTION
**Pull request by northwestwitch**
This PR adds a functionality or fixes a bug.
- Display who added a variant in a ClinVar submission  #4326

image

<br>

- Add events for ClinVar variant add/Remove

image

<br><br>

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
1. Try adding/removing a variant from ClinVar and make sure it works

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [x] code approved by DNI
- [x] tests executed by CR, DN

### CODE REVIEW

File: scout/constants/case_tags.py [hunk 1]
```diff
@@ -62,57 +62,59 @@
 CASE_STATUSES = ("prioritized", "inactive", "active", "solved", "archived", "ignored")
 
 VERBS_MAP = {
+    "acmg": "updated ACMG classification for",
```
Comment (northwestwitch) [inline scout/constants/case_tags.py:65]: Reordered them aphabetically

File: scout/constants/case_tags.py [hunk 2]
```diff
@@ -62,57 +62,59 @@
...
+    "add_phenotype": "added HPO term for",
+    "archive": "archived",
     "assign": "was assigned to",
-    "unassign": "was unassigned from",
-    "status": "updated the status for",
+    "beacon_add": "exported variants to the Beacon",
+    "beacon_remove": "removed variants from the Beacon",
+    "cancel_sanger": "cancelled sanger order for",
+    "cancer_tier": "updated cancer tier for",
+    "clinvar_add": "Added a variant to a ClinVar submission",
```
Comment (northwestwitch) [inline scout/constants/case_tags.py:75]: 2 new verbs

File: scout/adapter/mongo/clinvar.py [hunk 3]
```diff
@@ -257,15 +258,20 @@ def clinvar_submissions(self, institute_id):
                 )
 
                 # Loop over variants contained in a single ClinVar submission
-                for var_data_id in list(result["variant_data"]):
+                for var_info in submission["variant_data"]:
                     # get case_id from variant id (caseID_variant_ID)
-                    case_id = var_data_id.rsplit("_", 1)[0]
-                    CASE_CLINVAR_SUMBMISSION_PROJECTION = {"display_name": 1}
```
Comment (dnil) [inline scout/adapter/mongo/clinvar.py:263]: 😄

File: scout/adapter/mongo/clinvar.py [hunk 4]
```diff
@@ -437,3 +443,23 @@ def clinvar_cases(self, institute_id=None):
...
+        self, institute_id: str, case_id: str, variant_id: str
+    ) -> Optional[str]:
+        """Return the name of the user which added a specific variant to a submission."""
+        case_events_query = {
+            "verb": "clinvar_add",
+            "institute": institute_id,
+            "case": case_id,
+            "category": "variant",
+        }
+        projection = {"_id": 0, "user_name": 1, "link": 1}
```
Comment (dnil) [inline scout/adapter/mongo/clinvar.py:457]: 👍

**Review by dnil — APPROVED** [review]
Great stuff - this is quite useful. Stamped and approved, but I would prefer a move of the business part of this to corresponding controllers before merging.

File: scout/adapter/mongo/clinvar.py [hunk 5]
```diff
@@ -437,3 +443,23 @@ def clinvar_cases(self, institute_id=None):
...
+        clinvar_vars_for_case: pymongo.cursor.Cursor = self.event_collection.find(
+            case_events_query, projection
+        ).sort("created_at", -1)
+
+        for clinvar_var in clinvar_vars_for_case:
+            if variant_id in clinvar_var["link"]:
+                LOG.warning(clinvar_var)
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/adapter/mongo/clinvar.py:464]: Yes, I added it when I was checking why the latest test wasn't working! 😆

File: scout/server/blueprints/clinvar/views.py [hunk 6]
```diff
@@ -47,6 +47,22 @@ def clinvar_save(institute_id, case_name):
             "An open ClinVar submission was updated correctly with submitted data",
             "success",
         )
+        # Create user-related events
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/clinvar/views.py]: Done!

[the remark this answers is not in the document]
Comment (dnil) [reply scout/server/blueprints/clinvar/views.py]: Nice @northwestwitch - ahem, but check the inst and case - you likely don't need to fetch them twice!

[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/clinvar/views.py]: > Nice @northwestwitch - ahem, but check the inst and case - you likely don't need to fetch them twice!

You are right, that's what happens when you copy and paste! Thanks!

File: scout/server/blueprints/clinvar/views.py [hunk 7]
```diff
@@ -77,14 +93,34 @@ def clinvar_rename_casedata(submission, case, old_name):
...
+    if object_type == "variant_data":
+        variant_id: str = subm_variant_id.split("_")[-1]
+        variant_obj: dict = store.variant(document_id=variant_id)
+        institute_obj: dict = store.institute(institute_id=variant_obj["institute"])
+        case_obj: dict = store.case(case_id=variant_obj["case_id"])
+        for category in ["case", "variant"]:
+            store.create_event(
+                institute=institute_obj,
+                case=case_obj,
+                user=store.user(user_id=current_user._id),
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/clinvar/views.py]: Not in the same function, or?

File: scout/server/blueprints/clinvar/views.py [hunk 8]
```diff
@@ -77,14 +93,34 @@ def clinvar_rename_casedata(submission, case, old_name):
...
 
+    subm_variant_id: str = request.form.get("delete_object")
     store.delete_clinvar_object(
-        object_id=request.form.get("delete_object"),
+        object_id=subm_variant_id,
         object_type=object_type,
         submission_id=submission,
     )
+
+    # If variant itself is removed, register event
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/clinvar/views.py]: Fixed

File: scout/server/blueprints/clinvar/views.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dnil) [reply scout/server/blueprints/clinvar/views.py]: Its a small loop, so just twice atm. Your other function had a temp var for user..

[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/clinvar/views.py]: Ah OK, I'll put it outside, good point

**Pull request merged**